migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return nil, err
+	}
+	// have we migrated this?
+	if _, alreadyMigrated := m.cache[registerId.Owner]; alreadyMigrated {
+		// yes, return it as is
+		return []ledger.Payload{p}, nil
+	}
+	if !needsMigration(registerId) {
+		return []ledger.Payload{p}, nil
+	}
+
+	// we are going to migrate this
+	m.cache[registerId.Owner] = struct{}{}
+
+	if em, hasEM := m.ExceptionalMigrations[registerId.Owner]; hasEM {
+		m.logger.Info().
+			Err(err).
+			Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+			Msg("Using exceptional migration for address")
+		return em(p, m.logger)
+	}
+
+	return m.migrateValue(p)
+}
+
+func (m *MultipleContractMigration) migrateValue(p ledger.Payload) ([]ledger.Payload, error) {
+
+	// we don't need the the empty code register
+	value := p.Value
+	address := flow.BytesToAddress(p.Key.KeyParts[0].Value)
+	if len(value) == 0 {
+		return []ledger.Payload{}, nil
+	}
+
+	code := string(value)
+	program, err := parser2.ParseProgram(code)
+	if err != nil {
+		m.logger.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Cannot parse program at address")
+		return []ledger.Payload{}, err
+	}
+
+	if len(program.Declarations) == 0 {
+		// If there is no declarations. Only comments? was this legal before?
+		// error just in case
+		// alternative would be removing the register
+		m.logger.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("No declarations at address")
+		return []ledger.Payload{}, fmt.Errorf("no declarations at address %s", address.Hex())
+	}
+
+	if len(program.Declarations) > 2 {
+		m.logger.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("More than two declarations at address")
+		return []ledger.Payload{}, fmt.Errorf("more than two declarations at address %s", address.Hex())
+	}
+
+	if len(program.Declarations) == 1 {
+		// If there is one declaration move it to the new key
+		m.logger.Debug().
+			Str("address", address.Hex()).
+			Msg("Single contract or interface at address moved to new key")
+		p.Key = addNameToKey(p.Key, program.Declarations[0].DeclarationIdentifier().Identifier)
+		return []ledger.Payload{p}, nil
+	}
+
+	// We have two declarations. Due to the current rules one of them is an interface and one is a contract.
+	// the contract will need an import to the interface.
+	m.logger.Info().
+		Str("address", address.Hex()).
+		Msg("Two declarations on an address, splitting into two parts")
+
+	_, oneIsInterface := program.Declarations[0].(*ast.InterfaceDeclaration)
+	_, twoIsInterface := program.Declarations[1].(*ast.InterfaceDeclaration)
+	if oneIsInterface == twoIsInterface {
+		// declarations of same type! should not happen
+		m.logger.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Two declarations of the same type at address")
+		return []ledger.Payload{}, fmt.Errorf("wwo declarations of the same type at address %s", address.Hex())
```

Comment:
wwo -> two
I think we should panic as well here, there is a high chance we miss this and wipe the register